examples/metrics-advanced/src/main.rs

```diff
@@ -49,13 +49,14 @@ fn init_meter_provider() -> opentelemetry_sdk::metrics::SdkMeterProvider {
         .with_temporality(Temporality::Delta)
         .build();
 
+    let resource = Resource::builder()
+        .with_service_name("metrics-advanced-example")
+        .build();
+
     let reader = PeriodicReader::builder(exporter, runtime::Tokio).build();
     let provider = SdkMeterProvider::builder()
         .with_reader(reader)
-        .with_resource(Resource::new([KeyValue::new(
-            "service.name",
-            "metrics-advanced-example",
-        )]))
+        .with_resource(resource)
```

Comment:
1. If user never called with_resource.-  use SDK defaults.
2. with_resource(Resource::builder().with..with...build()) - Start with SDK defaults, and merge the rest.
3. with_resource(Resource::builder_empty().with..with...build()) - Start with empty, and merge the rest.
4. provider.append_resource(..) - merges with existing Resource(s)

^ Does this look the right approach @open-telemetry/rust-approvers please share your thoughts, this is important to get closure soon.

Reply:
After some more thinking into this, I believe there mainly two use-cases here:

1. Set resource to something very specific (remove default resources)
2. Append more resources to the existing ones

I believe a simple solution here could be:
- Introduce an API named `set_resource(resource)` which will replace the provider's resource with the value in the method argument. This would cover the 1st use case.
- Use the existing `with_resource(resource)` API (or rename it to `append_resource(resource)` to allow users to append/merge resources with the existing provider resource. The existing resource would either be the default ones or something else if the user called `set_resource(resource)`. This would cover the 2nd use case.

Reply:
Do we have the use-case where resource would be updated/merged in Provider after it's creation. Atleast, I don't see it.  The user needs to ensure that it creates the resource object by merging all the default/custom/required detectors before creating the Provider. And in that case, set_resource(), add_resource() and merge_resource() on Provider is not required. So,

These looks fine:
- If user never called with_resource.- use SDK defaults.

This is also fine:
- with_resource(Resource::builder().with..with...build()) - Start with SDK defaults, and merge the rest.
- with_resource(Resource::builder_empty().with..with...build()) - Start with empty, and merge the rest.

Or maybe another option could be:
`Resource::builder(start_with_default: bool) -> ResourceBuilder`
i.e,
- with_resource(Resource::builder(start_with_default: true).with..with...build()) - Start with SDK defaults, and merge the rest.
- with_resource(Resource::builder(start_with_default: false).with..with...build()) - Start with SDK defaults, and merge the rest.
And `ResourceBuilder::new()` and `ResourceBuilder::default()` can be made private/removed.

But _NOT_ this, as I don't see it's use-case:
- provider.append_resource(..) - merges with existing Resource(s). 

the user should instead do invoke
```rust
  let merged_resource = resource.merge(to_be_merged);
  let provider = Provider::builder().with_resource(merged_resource)... .buld();
```

Reply:
I like the idea of using defaults, and giving the option of overriding them with `with_resource(Resource::empty())`, im not sure if it would be redundant to also provide `Resource::builder_empty()`, but I also like that option.

I'm not partial to `Resource::builder(true)` because it doesn't feel idiomatic requiring an argument to toggle behavior with a bool.

Reply:
> I'm not partial to Resource::builder(true) because it doesn't feel idiomatic requiring an argument to toggle behavior with a bool.

Agree, this doesn't look idiomatic. 

> I like the idea of using defaults, and giving the option of overriding them with with_resource(Resource::empty()), im not sure if it would be redundant to also provide Resource::builder_empty(), but I also like that option.

I don't see `Resource::builder_empty()` redundant. It would be to used to start building the resource empty, how can we achieve this otherwise?

Reply:
The with_resource could be given a empty resource to override the default one

Reply:
> I like the idea of having builder_empty, it makes the option more obvious

Yes, I also prefer this. Also correct me if I am wrong, but I don't see use of `Resource::empty()`, `Resource::new_with_defaults()`, `Resource::new()`, `ResourceBuilder::new()`, `ResoureBuilder::defaults()` to be public. `Resource::builder()` and `Resource::builder_empty()` should be the only entry point to create resources.

Reply:
> Resource::builder() and Resource::builder_empty() should be the only entry point to create resources.

This is good. Let us remove all other public APIs other than this.


_provider.append_resource(..)_
Agreed to *not* offer this in 1.0. If we chose to add it later, it is an additive change only. (There were scenarios which required this, especially from .NET feedbacks... But its best to be conservative in 1.0 release).

Reply:
Thank you all for the feedback! I should be able to apply it tonight after work

Reply:
so we want to remove `Resource::empty()` from public api? This seems to be used in quite a few places, just want to confirm before I remove it

Reply:
pushed up most of the suggestions other than removing the empty from public api